File: tif_customization/tif_customization/page/school_analytics/school_analytics.py
import frappe
from frappe.utils import flt


@frappe.whitelist()
def get_school_analytics(school):
	try:
		if not school:
			return {"error": "School is required"}

		doc = frappe.db.get_value(
			"School",
			school,
			[
				"name",
				"school_name",
				"status",
				"territory",
				"city",
				"school_type",
				"books",
				"trainings",
				"tps",
				"qps",
				"cee",
				"no_of_students",
				"total_no_of_quranic_teachers",
			],
			as_dict=True,
		)
		if not doc:
			return {"error": "Invalid School"}

		school_name = doc.get("school_name")
		issued_map = _get_issued_books_map(school_name)
		required_map = _get_required_books_map(school)

		book_status = []
		for book_type in ["QAIDA", "GUIDE", "MQH"]:
			required = flt(required_map.get(book_type, 0))
			issued = flt(issued_map.get(book_type, 0))
			remaining = max(required - issued, 0)

			status = "Complete" if remaining <= 0 and required > 0 else "Pending"
			if required == 0 and issued > 0:
				status = "Issued"
			if required == 0 and issued == 0:
				status = "Pending"

			book_status.append(
				{
					"book_type": book_type,
					"required": required,
					"issued": issued,
					"remaining": remaining,
					"status": status,
				}
			)

		completion = _get_profile_completion(doc)
		monthly = _get_monthly_dispatch(school_name)

		total_required = sum(d["required"] for d in book_status)
		total_issued = sum(d["issued"] for d in book_status)
		total_remaining = sum(d["remaining"] for d in book_status)

		return {
			"school_details": doc,
			"kpis": {
				"students": flt(doc.get("no_of_students") or 0),
				"quranic_teachers": flt(doc.get("total_no_of_quranic_teachers") or 0),
				"total_required_books": total_required,
				"total_issued_books": total_issued,
				"total_remaining_books": total_remaining,
				"profile_completion": completion["percent"],
			},
			"completion": completion,
			"book_status": book_status,
			"charts": {
				"books": _build_books_chart(book_status),
				"completion": _build_completion_chart(completion),
				"monthly": _build_monthly_chart(monthly),
			},
		}
	except Exception as exc:
		frappe.log_error(frappe.get_traceback(), "School Analytics Error")
		return {"error": str(exc)}


def _get_issued_books_map(school_name):
	if not school_name:
		return {"QAIDA": 0, "GUIDE": 0, "MQH": 0}

	row = frappe.db.sql(
		"""
		SELECT
			SUM(
				CASE
					WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%QAIDA%%'
						OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%QAIDA%%'
					THEN COALESCE(dni.qty, 0) ELSE 0
				END
			) AS qaida_issued,
			SUM(
				CASE
					WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%GUIDE%%'
						OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%GUIDE%%'
					THEN COALESCE(dni.qty, 0) ELSE 0
				END
			) AS guide_issued,
			SUM(
				CASE
					WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%MQH%%'
						OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%MQH%%'
					THEN COALESCE(dni.qty, 0) ELSE 0
				END
			) AS mqh_issued
		FROM `tabDelivery Note` dn
		INNER JOIN `tabDelivery Note Item` dni ON dni.parent = dn.name
		WHERE dn.docstatus = 1
			AND IFNULL(dn.is_return, 0) = 0
			AND dn.customer_name = %(school_name)s
		""",
		{"school_name": school_name},
		as_dict=True,
	)

	first = (row and row[0]) or {}
	return {
		"QAIDA": flt(first.get("qaida_issued") or 0),
		"GUIDE": flt(first.get("guide_issued") or 0),
		"MQH": flt(first.get("mqh_issued") or 0),
	}


def _get_required_books_map(school):
	parent = frappe.db.get_value(
		"School",
		school,
		["no_of_students", "total_no_of_quranic_teachers"],
		as_dict=True,
	)
	students = flt((parent or {}).get("no_of_students") or 0)
	teachers = flt((parent or {}).get("total_no_of_quranic_teachers") or 0)

	curriculum_total = frappe.db.sql(
		"""
		SELECT SUM(no_of_student) AS total_students
		FROM `tabSchool Books Curriculum`
		WHERE parent = %(school)s
		""",
		{"school": school},
		as_dict=True,
	)[0]["total_students"]

	mqh_required = flt(curriculum_total or 0) or students

	return {"QAIDA": students, "GUIDE": teachers, "MQH": mqh_required}


def _get_profile_completion(doc):
	# Core profile fields that should ideally be filled for complete school profiling.
	fields = {
		"School Name": doc.get("school_name"),
		"Status": doc.get("status"),
		"Territory": doc.get("territory"),
		"City": doc.get("city"),
		"School Type": doc.get("school_type"),
		"Students": doc.get("no_of_students"),
		"Quranic Teachers": doc.get("total_no_of_quranic_teachers"),
		"Books": doc.get("books"),
		"Trainings": doc.get("trainings"),
		"TPS": doc.get("tps"),
		"QPS": doc.get("qps"),
		"CEE": doc.get("cee"),
	}

	done, remaining = [], []
	for label, value in fields.items():
		if value in [None, "", 0]:
			remaining.append(label)
		else:
			done.append(label)

	total = len(fields) or 1
	percent = round((len(done) / total) * 100, 1)

	return {
		"done_count": len(done),
		"total_count": total,
		"percent": percent,
		"remaining_fields": remaining,
	}


def _get_monthly_dispatch(school_name):
	if not school_name:
		return []

	rows = frappe.db.sql(
		"""
		SELECT
			DATE_FORMAT(dn.posting_date, '%%Y-%%m') AS month_key,
			SUM(
				CASE
					WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%MQH%%'
						OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%MQH%%'
					THEN COALESCE(dni.qty, 0) ELSE 0
				END
			) AS mqh_qty,
			SUM(
				CASE
					WHEN UPPER(COALESCE(dni.item_name, '')) LIKE '%%QAIDA%%'
						OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%QAIDA%%'
						OR UPPER(COALESCE(dni.item_name, '')) LIKE '%%GUIDE%%'
						OR UPPER(COALESCE(dni.item_code, '')) LIKE '%%GUIDE%%'
					THEN COALESCE(dni.qty, 0) ELSE 0
				END
			) AS qaida_guide_qty
		FROM `tabDelivery Note` dn
		INNER JOIN `tabDelivery Note Item` dni ON dni.parent = dn.name
		WHERE dn.docstatus = 1
			AND IFNULL(dn.is_return, 0) = 0
			AND dn.customer_name = %(school_name)s
			AND dn.posting_date >= DATE_SUB(CURDATE(), INTERVAL 12 MONTH)
		GROUP BY DATE_FORMAT(dn.posting_date, '%%Y-%%m')
		ORDER BY month_key
		""",
		{"school_name": school_name},
		as_dict=True,
	)
	return rows


def _build_books_chart(book_status):
	labels = [d["book_type"] for d in book_status]
	issued = [d["issued"] for d in book_status]
	remaining = [d["remaining"] for d in book_status]

	return {
		"labels": labels,
		"datasets": [
			{"label": "Issued", "data": issued, "backgroundColor": "#198754"},
			{"label": "Remaining", "data": remaining, "backgroundColor": "#dc3545"},
		],
	}


def _build_completion_chart(completion):
	done = completion.get("done_count", 0)
	total = completion.get("total_count", 0)
	pending = max(total - done, 0)
	return {
		"labels": ["Completed", "Pending"],
		"datasets": [
			{
				"data": [done, pending],
				"backgroundColor": ["#0d6efd", "#ced4da"],
			}
		],
	}


def _build_monthly_chart(rows):
	labels = [r.get("month_key") for r in rows]
	mqh = [flt(r.get("mqh_qty") or 0) for r in rows]
	qaida_guide = [flt(r.get("qaida_guide_qty") or 0) for r in rows]

	return {
		"labels": labels,
		"datasets": [
			{
				"label": "MQH Dispatch",
				"data": mqh,
				"borderColor": "#0d6efd",
				"backgroundColor": "rgba(13,110,253,0.2)",
				"fill": False,
			},
			{
				"label": "Qaida/Guide Dispatch",
				"data": qaida_guide,
				"borderColor": "#20c997",
				"backgroundColor": "rgba(32,201,151,0.2)",
				"fill": False,
			},
		],
	}

File: tif_customization/tif_customization/page/school_analytics/school_analytics.js
frappe.pages["school-analytics"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "School Analytics",
		single_column: true,
	});

	new SchoolAnalyticsPage(page).make();
};

class SchoolAnalyticsPage {
	constructor(page) {
		this.page = page;
		this.school_filter = null;
		this.charts = {};
	}

	make() {
		this.make_filters();
		this.make_layout();
		this.load_chartjs(() => this.load_data());
	}

	make_filters() {
		this.school_filter = this.page.add_field({
			label: __("School"),
			fieldtype: "Link",
			fieldname: "school",
			options: "School",
			reqd: 1,
			change: () => this.load_data(),
		});

		this.page.set_primary_action(__("Refresh"), () => this.load_data(), "refresh");
	}

	make_layout() {
		const html = `
			<div class="school-analytics-wrap" style="padding:16px;">
				<div id="sa-empty" class="text-muted" style="padding:16px; border:1px solid #d1d8dd; border-radius:8px; margin-bottom:16px;">
					Select a school to view analytics.
				</div>

				<div id="sa-content" style="display:none;">
					<div id="sa-kpis" class="row" style="margin-bottom:16px;"></div>

					<div class="row" style="margin-bottom:16px;">
						<div class="col-md-6">
							<div style="border:1px solid #d1d8dd; border-radius:8px; padding:12px; background:#fff;">
								<div style="font-weight:600; margin-bottom:8px;">Book Issue: Issued vs Remaining</div>
								<canvas id="sa-books-chart" height="210"></canvas>
							</div>
						</div>
						<div class="col-md-6">
							<div style="border:1px solid #d1d8dd; border-radius:8px; padding:12px; background:#fff;">
								<div style="font-weight:600; margin-bottom:8px;">School Profile Completion</div>
								<canvas id="sa-completion-chart" height="210"></canvas>
							</div>
						</div>
					</div>

					<div style="border:1px solid #d1d8dd; border-radius:8px; padding:12px; background:#fff; margin-bottom:16px;">
						<div style="font-weight:600; margin-bottom:8px;">Monthly Dispatch Trend</div>
						<canvas id="sa-monthly-chart" height="100"></canvas>
					</div>

					<div class="row">
						<div class="col-md-6">
							<div style="border:1px solid #d1d8dd; border-radius:8px; background:#fff; padding:12px; margin-bottom:16px;">
								<div style="font-weight:600; margin-bottom:8px;">School Details</div>
								<div id="sa-details"></div>
							</div>
						</div>
						<div class="col-md-6">
							<div style="border:1px solid #d1d8dd; border-radius:8px; background:#fff; padding:12px; margin-bottom:16px;">
								<div style="font-weight:600; margin-bottom:8px;">Remaining Profile Fields</div>
								<div id="sa-remaining-fields"></div>
							</div>
						</div>
					</div>

					<div style="border:1px solid #d1d8dd; border-radius:8px; background:#fff; padding:12px;">
						<div style="font-weight:600; margin-bottom:8px;">Book Issue Status</div>
						<div class="table-responsive">
							<table class="table table-bordered" style="margin-bottom:0;">
								<thead>
									<tr>
										<th>Book Type</th>
										<th class="text-right">Required</th>
										<th class="text-right">Issued</th>
										<th class="text-right">Remaining</th>
										<th>Status</th>
									</tr>
								</thead>
								<tbody id="sa-book-status-tbody"></tbody>
							</table>
						</div>
					</div>
				</div>
			</div>
		`;

		$(this.page.body).find(".school-analytics-wrap").remove();
		$(this.page.body).append(html);
	}

	load_chartjs(callback) {
		if (window.Chart) {
			callback();
			return;
		}

		frappe.require("https://cdn.jsdelivr.net/npm/chart.js@4.4.0/dist/chart.umd.min.js", () => {
			callback();
		});
	}

	load_data() {
		const school = this.school_filter ? this.school_filter.get_value() : "";
		if (!school) {
			$("#sa-content").hide();
			$("#sa-empty").show();
			return;
		}

		frappe.call({
			method: "tif_customization.tif_customization.page.school_analytics.school_analytics.get_school_analytics",
			args: { school },
			callback: (r) => {
				const data = r.message || {};
				if (data.error) {
					frappe.msgprint(__("Unable to load analytics: {0}", [data.error]));
					return;
				}
				$("#sa-empty").hide();
				$("#sa-content").show();
				this.render_all(data);
			},
			error: () => frappe.msgprint(__("Failed to load School Analytics.")),
		});
	}

	render_all(data) {
		this.render_kpis(data.kpis || {});
		this.render_details(data.school_details || {});
		this.render_remaining_fields(data.completion?.remaining_fields || []);
		this.render_book_table(data.book_status || []);
		this.render_charts(data.charts || {});
	}

	render_kpis(kpis) {
		const cards = [
			{ label: "Students", value: kpis.students || 0 },
			{ label: "Quranic Teachers", value: kpis.quranic_teachers || 0 },
			{ label: "Books Required", value: kpis.total_required_books || 0 },
			{ label: "Books Issued", value: kpis.total_issued_books || 0 },
			{ label: "Books Remaining", value: kpis.total_remaining_books || 0 },
			{ label: "Profile Completion", value: `${kpis.profile_completion || 0}%` },
		];

		const html = cards
			.map(
				(card) => `
				<div class="col-md-4 col-sm-6" style="margin-bottom:12px;">
					<div style="padding:12px; border:1px solid #d1d8dd; border-radius:8px; background:#f8f9fa;">
						<div class="text-muted small">${frappe.utils.escape_html(card.label)}</div>
						<div style="font-size:22px; font-weight:600;">${frappe.utils.escape_html(String(card.value))}</div>
					</div>
				</div>
			`
			)
			.join("");

		$("#sa-kpis").html(html);
	}

	render_details(details) {
		const rows = [
			["School", details.school_name || "-"],
			["Status", details.status || "-"],
			["Territory", details.territory || "-"],
			["City", details.city || "-"],
			["Type", details.school_type || "-"],
			["TPS", details.tps || "-"],
			["QPS", details.qps || "-"],
			["CEE", details.cee || "-"],
			["Trainings", details.trainings || "-"],
			["Books", details.books || "-"],
		];

		const html = `
			<table class="table table-sm table-bordered" style="margin-bottom:0;">
				<tbody>
					${rows
						.map(
							(row) =>
								`<tr><th style="width:35%;">${frappe.utils.escape_html(row[0])}</th><td>${frappe.utils.escape_html(
									String(row[1])
								)}</td></tr>`
						)
						.join("")}
				</tbody>
			</table>
		`;

		$("#sa-details").html(html);
	}

	render_remaining_fields(fields) {
		if (!fields.length) {
			$("#sa-remaining-fields").html('<div class="text-success">All required profile fields are complete.</div>');
			return;
		}

		const html = `<ul style="margin-bottom:0; padding-left:18px;">${fields
			.map((f) => `<li>${frappe.utils.escape_html(f)}</li>`)
			.join("")}</ul>`;
		$("#sa-remaining-fields").html(html);
	}

	render_book_table(rows) {
		if (!rows.length) {
			$("#sa-book-status-tbody").html('<tr><td colspan="5" class="text-muted text-center">No data</td></tr>');
			return;
		}

		const html = rows
			.map((row) => {
				const status_class = row.status === "Complete" ? "text-success" : "text-danger";
				return `
					<tr>
						<td>${frappe.utils.escape_html(row.book_type || "-")}</td>
						<td class="text-right">${this.fnum(row.required)}</td>
						<td class="text-right">${this.fnum(row.issued)}</td>
						<td class="text-right">${this.fnum(row.remaining)}</td>
						<td class="${status_class}">${frappe.utils.escape_html(row.status || "-")}</td>
					</tr>
				`;
			})
			.join("");

		$("#sa-book-status-tbody").html(html);
	}

	render_charts(charts) {
		if (!window.Chart) return;

		this.destroy_chart("books");
		this.destroy_chart("completion");
		this.destroy_chart("monthly");

		const booksCtx = document.getElementById("sa-books-chart");
		const completionCtx = document.getElementById("sa-completion-chart");
		const monthlyCtx = document.getElementById("sa-monthly-chart");

		if (booksCtx && charts.books) {
			this.charts.books = new Chart(booksCtx, {
				type: "bar",
				data: charts.books,
				options: {
					responsive: true,
					plugins: { legend: { position: "top" } },
				},
			});
		}

		if (completionCtx && charts.completion) {
			this.charts.completion = new Chart(completionCtx, {
				type: "doughnut",
				data: charts.completion,
				options: {
					responsive: true,
					plugins: { legend: { position: "bottom" } },
				},
			});
		}

		if (monthlyCtx && charts.monthly) {
			this.charts.monthly = new Chart(monthlyCtx, {
				type: "line",
				data: charts.monthly,
				options: {
					responsive: true,
					plugins: { legend: { position: "top" } },
					elements: { line: { tension: 0.3 } },
				},
			});
		}
	}

	destroy_chart(key) {
		if (this.charts[key]) {
			this.charts[key].destroy();
			this.charts[key] = null;
		}
	}

	fnum(v) {
		return format_number(v || 0, null, 0);
	}
}
